gitpod-io/gitpod
# Implement Project-level environment variables

## Description


Introduce Project-level environment variables:

- [x] [dashboard] Unify `project`-getting code into a `ProjectContext` helper
- [x] Implement a Variables page in Project Settings
- [x] Implement Project-level variables in `server` and `gitpod-db`
- [x] Expose Project-level variables in prebuilds
- [x] Expose Project-level variables in workspaces...
- [x] ...but also hiding certain variables in workspaces
- [ ] ...and censoring them in logs too (image builds, prebuild logs, etc. in order to mitigate accidental leaks of secrets) (EDIT: follow-up)
- [x] ~~Drive-by: Export `server`'s `censor` function via `gitpod-protocol` (in order to use it in `gitpod-db`)~~ (EDIT: Commit removed, censor code copy/pasted and specialized)
- [x] ~~Drive-by: Always use the actual function name (`arguments.callee.name`) in `checkUser` and `checkAndBlockUser` (instead of re-typing/copy-pasting the function name and sometimes making mistakes)~~ (EDIT: Commit removed again, `arguments.callee` is [deprecated](https://stackoverflow.com/questions/38435450/get-current-function-name-in-strict-mode) 🙄)

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/4456

## How to test


1. Log in to https://jx-project-variables.staging.gitpod-dev.com/projects
1. In your User Settings, create an environment variable called `USER` with value `user` and pattern `*/*`
1. Next, create a project
1. Navigate to Project Settings > Variables, and create a new variable (e.g. `PROJECT` with value `project`, visible in both Prebuilds and Workspaces)
1. In the Project repository, create a new branch containing a `.gitpod.yml` file that looks like this:

```yml
tasks:
  - init: echo "USER: $USER, PROJECT: $PROJECT"
    command: echo "USER: $USER, PROJECT: $PROJECT"
```

- Verify that, in the triggered prebuild (from the new branch created in 5.), only the `PROJECT` variable is defined (`USER` should still be undefined in the prebuild)
- When you open a workspace for that new branch, both `USER` and `PROJECT` should be defined (not in the printed Prebuild logs, but in the interactive command & also in live Terminals)

~~Open question: If you open a workspace for the new branch as a new user who doesn't have access to the Project (i.e. the repository is public, and you open a workspace for it, which gives you a prebuild, but you have no access to the Gitpod Project) -- should you be able to see the Project variables? (Prebuild/non-prebuild consistency says "yes", secrets say "no", maybe this should be configurable for each variable, see https://github.com/gitpod-io/gitpod/issues/4456#issuecomment-994953171)~~ (EDIT: Visibility is now configurable)

## Release Notes

```release-note
Introduce Project-level environment variables
```

## Documentation



/uncc
